python/ray/serve/http_proxy.py

```diff
@@ -1,23 +1,82 @@
 import asyncio
 import socket
-from typing import List
+from typing import List, Dict, Tuple
 
 import uvicorn
 import starlette.responses
+import starlette.routing
 
 import ray
 from ray.exceptions import RayTaskError
+from ray.serve.common import EndpointTag
 from ray.serve.constants import LongPollKey
 from ray.util import metrics
 from ray.serve.utils import _get_logger
 from ray.serve.http_util import Response, build_starlette_request
 from ray.serve.long_poll import LongPollAsyncClient
-from ray.serve.router import Router
 from ray.serve.handle import DEFAULT
 
 logger = _get_logger()
 
 
+class ServeStarletteEndpoint:
+    """Wraps the given Serve endpoint in a Starlette endpoint.
+
+    Implements the ASGI protocol.  Constructs a Starlette endpoint for use by
+    a Starlette app or Starlette Router which calls the given Serve endpoint
+    using the given Serve client.
+
+    Usage:
+        route = starlette.routing.Route(
+                "/api",
+                ServeStarletteEndpoint(self.client, endpoint_tag),
+                methods=methods)
+        app = starlette.applications.Starlette(routes=[route])
+    """
+
+    def __init__(self, client, endpoint_tag: EndpointTag):
+        self.client = client
+        self.endpoint_tag = endpoint_tag
+
+    async def __call__(self, scope, receive, send):
+        http_body_bytes = await self.receive_http_body(scope, receive, send)
+
+        headers = {k.decode(): v.decode() for k, v in scope["headers"]}
+
+        handle = self.client.get_handle(
+            self.endpoint_tag, sync=False).options(
+                method_name=headers.get("X-SERVE-CALL-METHOD".lower(),
+                                        DEFAULT.VALUE),
+                shard_key=headers.get("X-SERVE-SHARD-KEY".lower(),
+                                      DEFAULT.VALUE),
+                http_method=scope["method"].upper(),
```

Comment:
Should we be caching the handle instead of getting it every time a request comes in? Shouldn't be a big performance difference but logically it seems to make more sense too.

Reply:
Yes, I think the handles are cached in the client--is this sufficient?

Reply:
they are not cached in the client anymore. i think we should cache them here. preferable way to do it will be a property method decorated by functools.lru_cache

Reply:
hmm we shouldn't need to do lru_cache here because we know exactly the endpoints that are in use, right? (those that have routes associated with them)

Reply:
Since this block of code is already scoped to a specific endpoint, only one handle needs to be cached here.  I've added this change.  

`handle.options()` is still called each time a request comes in, because `headers` and `method` can change.  `handle.options()` constructs a new handle object, but I think it should still be faster than client.get_handle().  Is this true, or is it not actually significantly faster?

The alternative would be to do an LRU cache for each endpoint with multiple handles for each endpoint, with the handles keyed by the `headers` dict and `method`.  This requires some small hack to make the `headers` dict hashable.